### PR TITLE
resource_manager: report ru metrics to PD

### PR DESCRIPTION
**Pull request by HuSharp**
### What is changed and how it works?

Issue Number: Ref #14900
wait for #15100 merged

What's Changed:

- Support for the `upload metrics` interface on the PD client, which uses the `acquire_token_buckets` interface in the PD server.
- Change io consumption to distinguish between reads and writes for ru cost billing
- Uploading metrics requires first getting the resource manager configuration from etcd in the PD
- Calculate current ru cost by subtracting last io&cpu subsection from current io&cpu
- When the limiter is restarted, the version will change, when the version changes, we need to recalculate the ru cost
- Add mocker pd related function to check whether upload to pd

```commit-message
```

image

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

```release-note
Support for the `upload metrics` interface on the PD client, which uses the `acquire_token_buckets` interface in the PD server
```

### CODE REVIEW

File: components/resource_control/src/resource_limiter.rs [hunk 1]
```diff
@@ -44,10 +45,10 @@ impl ResourceLimiter {
         }
     }
 
-    pub fn consume(&self, cpu_time: Duration, io_bytes: u64) -> Duration {
+    pub fn consume(&self, cpu_time: Duration, io_bytes: IoBytes) -> Duration {
         let cpu_dur =
             self.limiters[ResourceType::Cpu as usize].consume(cpu_time.as_micros() as u64);
-        let io_dur = self.limiters[ResourceType::Io as usize].consume(io_bytes);
+        let io_dur = self.limiters[ResourceType::Io as usize].consume_io(io_bytes);
```
Comment (HuSharp) [inline components/resource_control/src/resource_limiter.rs:51]: @glorv PTAL for io consume, thx!

File: components/resource_control/src/service.rs [hunk 2]
```diff
@@ -139,6 +160,178 @@ impl ResourceManagerService {
...
+                        .get_limiter(ResourceType::Cpu)
+                        .get_statistics()
+                        .total_consumed as f64;
+                    let io_consumed = g.limiter.get_limiter(ResourceType::Io).get_statistics();
+                    let read_bytes_consumed = io_consumed.read_consumed as f64;
+                    let write_bytes_consumed = io_consumed.write_consumed as f64;
+
+                    let read_total = config.cpu_ms_cost * cpu_consumed
+                        + config.read_cost_per_byte * read_bytes_consumed
+                        + config.read_per_batch_base_cost * DEFAULT_AVG_BATCH_PROPORTION;
```
Comment (glorv) [inline components/resource_control/src/service.rs:247]: If you want to count `read/write_per_batch_base_cost`, you should count them either in the client or add a separate field to count the request count, add this default value makes no sense.

File: components/resource_control/src/resource_limiter.rs (hunk 1 above)
Comment (glorv) [reply components/resource_control/src/resource_limiter.rs:51]: LGTM

File: components/resource_control/src/service.rs [hunk 3]
```diff
@@ -139,6 +155,163 @@ impl ResourceManagerService {
...
+    read_base_cost: f64,
+    #[serde(rename = "read-cost-per-byte")]
+    read_cost_per_byte: f64,
+    #[serde(rename = "write-base-cost")]
+    write_base_cost: f64,
+    #[serde(rename = "write-cost-per-byte")]
+    write_cost_per_byte: f64,
+    #[serde(rename = "read-cpu-ms-cost")]
+    cpu_ms_cost: f64,
+}
```
Comment (glorv) [inline components/resource_control/src/service.rs:294]:
```suggestion
#[serde(rename_all = "kebab-case")]
struct RequestUnitConfig {
    read_base_cost: f64,
    read_cost_per_byte: f64,
    write_base_cost: f64,
    write_cost_per_byte: f64,
    read_cpu_ms_cost: f64,
}
```

File: components/resource_control/src/service.rs [hunk 4]
```diff
@@ -139,6 +155,163 @@ impl ResourceManagerService {
...
+        let config = self.load_controller_config().await.unwrap_or_default();
+
+        loop {
+            let background_groups: Vec<_> = self
+                .manager
+                .resource_groups
+                .iter()
+                .filter_map(|kv| {
+                    let g = kv.value();
+                    g.limiter.as_ref().map(|limiter| GroupStats {
```
Comment (glorv) [inline components/resource_control/src/service.rs:204]: Why reuse the GroupStats struct, you only depends on the GroupStatistics field here, better define a separated struct here.

File: components/resource_control/src/service.rs [hunk 5]
```diff
@@ -139,6 +155,163 @@ impl ResourceManagerService {
...
+
+            let mut req = TokenBucketsRequest::default();
+            let all_reqs = req.mut_requests();
+            background_groups.iter().for_each(|g| {
+                let mut req = TokenBucketRequest::default();
+                req.set_resource_group_name(g.name.clone());
+                req.set_is_background(true);
+                // update statistics.
+                let report_consumption = req.mut_consumption_since_last_request();
+                if let Some(last_consumption) = last_consumption_map.get_mut(g.name.as_str()) {
```
Comment (glorv) [inline components/resource_control/src/service.rs:229]: I suggest cache the GroupStatistics here and calculate the delta value first then you needn't do any `(cur - last).max(0)` in the following code. And you can easily check if the delta value is zero, then you needn't report this request.

File: components/resource_control/src/service.rs (hunk 3 above)
Comment (HuSharp) [reply components/resource_control/src/service.rs:294]: get it.

File: components/resource_control/src/service.rs (hunk 5 above)
Comment (HuSharp) [reply components/resource_control/src/service.rs]: done.

File: components/resource_control/src/service.rs (hunk 4 above)
Comment (HuSharp) [reply components/resource_control/src/service.rs]: Yes! fixed it as `UploadStatistic`, PTAL, thx!

File: components/resource_control/src/resource_limiter.rs [hunk 6]
```diff
@@ -94,10 +102,12 @@ impl QuotaLimiter {
         self.limiter.set_speed_limit(limit);
     }
 
-    fn get_statistics(&self) -> (u64, u64) {
+    pub(crate) fn get_statistics(&self) -> (u64, u64, u64, u64) {
```
Comment (glorv) [inline components/resource_control/src/resource_limiter.rs:105]: not need to be pub, should depend on ResourceLimiter::get_limit_statistics instead

File: components/resource_control/src/resource_limiter.rs [hunk 7]
```diff
@@ -112,13 +122,32 @@ impl QuotaLimiter {
         }
         dur
     }
+
+    #[allow(dead_code)]
```
Comment (glorv) [inline components/resource_control/src/resource_limiter.rs:126]: remove useless attr

File: components/resource_control/src/service.rs [hunk 8]
```diff
@@ -139,6 +151,180 @@ impl ResourceManagerService {
             }
         }
     }
+
+    async fn load_controller_config(&self) -> Result<RequestUnitConfig, ()> {
```
Comment (glorv) [inline components/resource_control/src/service.rs:155]: I don't see when this function will return Err, remove the Result wrapper type

File: components/resource_control/src/service.rs [hunk 9]
```diff
@@ -139,6 +151,180 @@ impl ResourceManagerService {
...
+
+                        let read_total = (config.read_cpu_ms_cost * cpu_consumed as f64
+                            + config.read_cost_per_byte * read_bytes_consumed as f64)
+                            as u64;
+                        let write_total =
+                            (config.write_cost_per_byte * write_bytes_consumed as f64) as u64;
+
+                        (
+                            g.group.name.clone(),
+                            UploadStatistic {
```
Comment (glorv) [inline components/resource_control/src/service.rs:216]: should keep the version field to handle the case that the statistics data is reset when the quota limiter is recreated.

File: components/resource_control/src/service.rs [hunk 10]
```diff
@@ -139,6 +151,180 @@ impl ResourceManagerService {
...
+struct RequestUnitConfig {
+    read_base_cost: f64,
+    read_cost_per_byte: f64,
+    write_base_cost: f64,
+    write_cost_per_byte: f64,
+    read_cpu_ms_cost: f64,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct ControllerConfig {
```
Comment (glorv) [inline components/resource_control/src/service.rs:288]: ditto

File: components/resource_control/src/worker.rs [hunk 11]
```diff
@@ -277,12 +277,12 @@ impl<R: ResourceStatsProvider> GroupQuotaAdjustWorker<R> {
     }
 }
 
-struct GroupStats {
-    name: String,
-    limiter: Arc<ResourceLimiter>,
-    ru_quota: f64,
-    stats_per_sec: GroupStatistics,
-    expect_cost_rate: f64,
+pub(crate) struct GroupStats {
```
Comment (glorv) [inline components/resource_control/src/worker.rs:280]: rollback this change

File: components/resource_control/src/service.rs (hunk 8 above)
Comment (HuSharp) [reply components/resource_control/src/service.rs]: Okay, it was modified to option. I think when meeting error needs to fallback to the default config other than check it regularly.

File: components/resource_control/src/service.rs (hunk 9 above)
Comment (HuSharp) [reply components/resource_control/src/service.rs:197]: Done. check io and cpu version.

File: components/server/src/server.rs [hunk 12]
```diff
@@ -341,24 +341,28 @@ where
 
         let resource_manager = if config.resource_control.enabled {
             let mgr = Arc::new(ResourceGroupManager::default());
-            let mut resource_mgr_service =
-                ResourceManagerService::new(mgr.clone(), pd_client.clone());
+            let resource_mgr_service = ResourceManagerService::new(mgr.clone(), pd_client.clone());
```
Comment (Connor1996) [inline components/server/src/server.rs:344]: How about extracting the background tasks spawning to resource group manager, so we don't need to write it twice

File: components/resource_control/src/resource_group.rs [hunk 13]
```diff
@@ -464,7 +464,7 @@ impl ResourceController {
 
     pub fn update_min_virtual_time(&self) {
         let start = Instant::now_coarse();
-        let mut min_vt = u64::MAX;
+        let mut min_vt = u64::MAX - 100_000;
```
Comment (Connor1996) [inline components/resource_control/src/resource_group.rs:467]: why

Comment (HuSharp) [reply components/resource_control/src/resource_group.rs:467]: Oh, just for test. Seems like the compiler cannot recognize there logic and reports an error `attempt to add with overflow` at `min_vt + 100_000`.
I will change it to `min_vt.wrapping_add(100_000)` to make it happy.

File: components/server/src/server.rs (hunk 12 above)
Comment (HuSharp) [reply components/server/src/server.rs]: done.

File: components/resource_control/src/lib.rs [hunk 14]
```diff
@@ -40,3 +45,33 @@ impl Default for Config {
         Self { enabled: true }
     }
 }
+
+pub fn start(
```
Comment (HuSharp) [inline components/resource_control/src/lib.rs:49]: @glorv @Connor1996 pls check here, thx!

File: components/resource_control/src/resource_group.rs [hunk 15]
```diff
@@ -477,7 +477,7 @@ impl ResourceController {
 
         // TODO: use different threshold for different resource type
         // needn't do update if the virtual different is less than 100ms/100KB.
-        if min_vt + 100_000 >= max_vt && max_vt < RESET_VT_THRESHOLD {
+        if min_vt.wrapping_add(100_000) >= max_vt && max_vt < RESET_VT_THRESHOLD {
```
Comment (HuSharp) [inline components/resource_control/src/resource_group.rs:480]: FYI, compiler thinks here maybe cause `attempt to add with overflow`

Comment (glorv) [reply components/resource_control/src/resource_group.rs:480]: if min_vt >= max_vt.saturating_sub(1000)

File: components/resource_control/src/service.rs [hunk 16]
```diff
@@ -192,6 +398,24 @@ pub mod tests {
         .unwrap();
     }
 
+    fn store_controller_config(pd_client: Arc<RpcClient>, config: ControllerConfig) {
```
[the remark this answers is not in the document]
Comment (HuSharp) [reply components/resource_control/src/service.rs:401]: Oh, just want to mock server store config in test.

File: components/resource_control/src/service.rs [hunk 17]
```diff
@@ -139,20 +152,213 @@ impl ResourceManagerService {
...
+                None
+            }
+        }
+    }
+
+    // upload ru metrics periodically.
+    pub async fn upload_ru_metrics(&self) {
+        let mut last_group_statistics_map: HashMap<String, UploadStatistic> = HashMap::new();
+        // load controller config firstly.
+        let config = self.load_controller_config().await.unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (HuSharp) [reply components/resource_control/src/service.rs:186]: Okay, There is keeping consistent with PD client. I will change pd to retry as well.

[the remark this answers is not in the document]
Comment (glorv) [reply components/resource_control/src/service.rs:205]: Why still `unwrap_or_default`? You should ensure fetch the *right* config from pd before starting the ru report. I suggest do not do any retry in the `load_controller_config` func but just return the original error and handle the error here with an explicit retry loop.

**Review by glorv — COMMENTED** [review]
rest LGTM

File: components/resource_control/src/service.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (HuSharp) [reply components/resource_control/src/service.rs:205]: Previously keep it only when retrying for long time... well, you are right. There must get the right config.

File: components/resource_control/src/service.rs [hunk 18]
```diff
@@ -139,20 +152,215 @@ impl ResourceManagerService {
...
+            }
+
+            let mut req = TokenBucketsRequest::default();
+            let all_reqs = req.mut_requests();
+            for (name, statistic) in background_groups.into_iter() {
+                // check if the entry exists in the map or create a new one with u64::MAX
+                // version to avoid losing the first upload.
+                let last_stats = last_group_statistics_map
+                    .entry(name.clone())
+                    .or_insert_with(|| UploadStatistic::new(u64::MAX));
```
Comment (glorv) [inline components/resource_control/src/service.rs:243]: why explicitly set u64::MAX here, I think the default value 0 should be enough.

Comment (HuSharp) [reply components/resource_control/src/service.rs:243]: This is a flag to check if it is the first time it is stored in `last_group_statistics_map`... Let me see if there is a better way to distinguish between `equal` and `new insert`.

Comment (HuSharp) [reply components/resource_control/src/service.rs]: fixed it by checking the existence and version, PTAL again! thx! @glorv

File: components/resource_control/src/service.rs [hunk 19]
```diff
@@ -278,22 +495,107 @@ pub mod tests {
...
+            .unwrap();
+        new_bg_limiter.consume(
+            Duration::from_secs(5),
+            IoBytes {
+                read: 2000,
+                write: 2000,
+            },
+        );
+        // Wait for upload ru metrics.
+        std::thread::sleep(Duration::from_millis(100));
```
Comment (glorv) [inline components/resource_control/src/service.rs:597]: How can this case ensure the metrics are reported to pd, I don't see the test point of this case

File: components/test_pd/src/mocker/mod.rs [hunk 20]
```diff
@@ -216,4 +220,11 @@ pub trait PdMocker {
     fn get_operator(&self, _: &GetOperatorRequest) -> Option<Result<GetOperatorResponse>> {
         None
     }
+
+    fn upload_ru_metrics(&self, req: &TokenBucketsRequest) -> Option<Result<TokenBucketsResponse>> {
+        req.get_requests().iter().for_each(|r| {
+            assert_eq!(r.get_is_background(), true);
```
Comment (HuSharp) [inline components/test_pd/src/mocker/mod.rs:226]: @glorv FIY, just check the label for req...

Comment (glorv) [reply components/test_pd/src/mocker/mod.rs:226]: But how can you ensure if this method is even called?

Comment (HuSharp) [reply components/test_pd/src/mocker/mod.rs:226]: - Change `upload_duration` to 10ms in cfg and wait 100 ms here to make sure it does.
- The `new_test_server_and_client` ensures that we pass the simulation pd.

**Review by nolouch — COMMENTED** [review]
rest lgtm

File: components/pd_client/src/client.rs [hunk 21]
```diff
@@ -1178,6 +1179,52 @@ impl PdClient for RpcClient {
             .request(req, executor, LEADER_CHANGE_RETRY)
             .execute()
     }
+
+    fn upload_ru_metrics(&self, req: TokenBucketsRequest) -> PdFuture<()> {
```
[the remark this answers is not in the document]
Comment (HuSharp) [reply components/pd_client/src/client.rs]: done.

**Review by nolouch — APPROVED** [review]

**Pull request merged**